python/ray/rllib/agents/marwil/marwil_policy_graph.py

```diff
@@ -0,0 +1,167 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from gym.spaces import Discrete
+import numpy as np
+import tensorflow as tf
+import tensorflow.contrib.layers as layers
+
+import ray
+from ray.rllib.models import ModelCatalog
+from ray.rllib.evaluation.postprocessing import compute_advantages
+from ray.rllib.evaluation.sample_batch import SampleBatch
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.error import UnsupportedSpaceException
+from ray.rllib.evaluation.policy_graph import PolicyGraph
+from ray.rllib.evaluation.tf_policy_graph import TFPolicyGraph
+from ray.rllib.agents.dqn.dqn_policy_graph import _scope_vars
+
+P_SCOPE = "p_func"
+V_SCOPE = "v_func"
+
+
+class ValueLoss(object):
+    def __init__(self, state_values, cummulative_rewards):
+        self.loss = 0.5 * tf.reduce_mean(
+            tf.square(state_values - cummulative_rewards))
+
+
+class ReweightedImitationLoss(object):
+    def __init__(self, state_values, cummulative_rewards, logits, actions,
+                 action_space, beta):
+        ma_adv_norm = tf.get_variable(
+            name="moving_average_of_advantage_norm",
+            dtype=tf.float32,
+            initializer=100.0,
+            trainable=False)
+        # advantage estimation
+        adv = cummulative_rewards - state_values
+        # update averaged advantage norm
+        update_adv_norm = tf.assign_add(
+            ref=ma_adv_norm,
+            value=1e-6 * (tf.reduce_mean(tf.square(adv)) - ma_adv_norm))
+
+        # exponentially weighted advantages
+        with tf.control_dependencies([update_adv_norm]):
+            exp_advs = tf.exp(
+                beta * tf.divide(adv, 1e-8 + tf.sqrt(ma_adv_norm)))
+
+        # log\pi_\theta(a|s)
+        dist_cls, _ = ModelCatalog.get_action_dist(action_space, {})
+        action_dist = dist_cls(logits)
+        logprobs = action_dist.logp(actions)
+
+        self.loss = -1.0 * tf.reduce_mean(tf.stop_gradient(adv) * logprobs)
```

Comment:
Shouldn't this be against `exp_advs` not `adv`?

Reply:
oh...my fault. Thanks!

Reply:
It very critical when the historical data include trajectories generated by different policies in which case the advantage weights for a certain (s, a) pair may vary a lot and exponential op makes them non-negative. I am very sorry. Thank you!